chunky/src/java/se/llbit/chunky/entity/BeaconBeam.java

```diff
@@ -61,27 +67,135 @@ public BeaconBeam(JsonObject json) {
     this.scale = json.get("scale").asDouble(1);
     this.height = json.get("height").asInt(256);
     this.pose = json.get("pose").object();
+
+    JsonObject materialsList = json.get("beamMaterials").object();
+    for (JsonMember obj : materialsList.members) {
+      BeaconBeamMaterial mat = new BeaconBeamMaterial(BeaconBeamMaterial.DEFAULT_COLOR);
+      mat.loadMaterialProperties(obj.value.object());
+      materials.put(Integer.parseInt(obj.name), mat);
+    }
   }
 
   @Override
-  public Collection<Primitive> primitives(Vector3 offset) {
-    Vector3 allPose = JsonUtil.vec3FromJsonArray(this.pose.get("all"));
-    return primitives(Transform.NONE
-        .translate(-0.5, -0.5, -0.5)
-        .scale(scale)
-        .rotateY(allPose.y) // Ignore x and z rotation because it breaks the beam into segments
-        .translate(0.5, 0.5, 0.5)
-        .translate(position.x + offset.x, position.y + offset.y, position.z + offset.z));
+  public void loadDataFromOctree(Octree octree, BlockPalette palette, Vector3i origin) {
+    int firstColor = BeaconBeamMaterial.DEFAULT_COLOR;
+    boolean foundFirst = false;
+    this.materials.put(0, new BeaconBeamMaterial(BeaconBeamMaterial.DEFAULT_COLOR));
+
+    //Start i at 1 so the first beacon is not checked.
+    //Stop i at 256 even if the beam is taller because the Octree will wrap the coordinates.
+    for (int i = 1; i < height && (i + position.y) < 256; i++) {
```

Comment:
This will not work for smaller octrees that use y cut-off or bigger octrees (1.17 world height or future cubic chunks support). You'll have to get the maximum y value from the octree somehow.

Reply:
Changed to check the Octree.isInside(...) method instead of using 256.